File: backend/app/api/pipeline.py
"""Stateless extraction pipeline: given the 3 uploaded files (paths already on
disk) and a couple of options, run the whole DXF/IFC extraction once and
return everything the caller needs. No server-side session is kept — the
caller (the API layer) owns the temp files for the duration of one request
and the response carries the full result."""

from __future__ import annotations

from app.extraction.axis_reference import AxisReference, build_axis_reference_from_dxf, build_axis_reference_from_ifc
from app.extraction.dxf_extractor import extract_dxf_state
from app.extraction.ifc_extractor import extract_ifc_state
from app.models.domain import Band, SourceFormat, StateKind, WidthSample


def _extract_state(path: str, fmt: SourceFormat, state: StateKind, axis: AxisReference, gabarit: str, dxf_step_m: float):
    if fmt == SourceFormat.DXF:
        bands, samples, mode = extract_dxf_state(path, state, axis, gabarit=gabarit, step_m=dxf_step_m)
        return bands, samples, mode
    bands, samples = extract_ifc_state(path, state, axis)
    return bands, samples, "ifc"


def run_extraction(
    axis_path: str,
    axis_fmt: SourceFormat,
    existant_path: str,
    existant_fmt: SourceFormat,
    projet_path: str,
    projet_fmt: SourceFormat,
    gabarit: str,
    dxf_step_m: float,
) -> tuple[AxisReference, list[Band], list[WidthSample], str, str]:
    if axis_fmt == SourceFormat.DXF:
        axis = build_axis_reference_from_dxf(axis_path)
    else:
        axis = build_axis_reference_from_ifc(axis_path)

    existant_bands, existant_samples, existant_mode = _extract_state(
        existant_path, existant_fmt, StateKind.EXISTANT, axis, gabarit, dxf_step_m
    )
    projet_bands, projet_samples, projet_mode = _extract_state(
        projet_path, projet_fmt, StateKind.PROJET, axis, gabarit, dxf_step_m
    )

    bands = existant_bands + projet_bands
    samples = existant_samples + projet_samples
    return axis, bands, samples, existant_mode, projet_mode

File: frontend/src/engine/pipeline.ts
/** Client-side equivalent of backend/app/api/pipeline.py: detects each
 * file's format by extension and runs the matching extractor, all three
 * files sharing one AxisReference. Runs inside the Web Worker (see
 * worker/extraction.worker.ts) so parsing never blocks the UI thread. */

import type { AxisReference } from "./axisReference";
import { buildAxisReferenceFromDxfContent } from "./dxf/axisReferenceDxf";
import { extractDxfState } from "./dxf/dxfExtractor";
import { buildAxisReferenceFromIfcModel } from "./ifc/axisReferenceIfc";
import { extractIfcState } from "./ifc/ifcExtractor";
import { getIfcApi, openModel } from "./ifc/webIfcClient";
import type { Point } from "./geometry";
import type { Band, StateKind, WidthSample } from "../types/domain";
import { getLog, resetLog, setLogContext, type LogEntry } from "./log";

export type SourceFormat = "dxf" | "ifc";

export function detectFormat(filename: string): SourceFormat {
  const ext = filename.toLowerCase().split(".").pop();
  if (ext === "dxf") return "dxf";
  if (ext === "ifc") return "ifc";
  throw new Error(`Format de fichier non supporté : .${ext} (attendu .dxf ou .ifc)`);
}

export interface FileInput {
  filename: string;
  format: SourceFormat;
  text?: string; // DXF: decoded text content
  bytes?: Uint8Array; // IFC: raw bytes
}

export interface ExtractionResult {
  axisConfidence: string;
  existantMode: string;
  projetMode: string;
  bands: Band[];
  samples: WidthSample[];
  /** The shared reference axis, in true plan (x, y) — drawn as its own
   * layer in the DXF export so the plan-view reconstruction has a spatial
   * reference, the same way it does in the source DXF/IFC files. */
  axisPoints: Point[];
  /** Notices raised while parsing the three files about anything that had
   * to be guessed or was missing outright (no real PK markers, a fallback
   * axis method, geometry excluded as decorative, an unmeasurable profile,
   * …) — see log.ts. */
  log: LogEntry[];
}

export async function runExtraction(
  axesProfils: FileInput,
  existant: FileInput,
  projet: FileInput,
  gabarit: string,
  dxfStepM: number | null,
  wasmBaseUrl: string,
): Promise<ExtractionResult> {
  resetLog();
  const needsIfc = [axesProfils, existant, projet].some((f) => f.format === "ifc");
  const api = needsIfc ? await getIfcApi(wasmBaseUrl) : null;
  const openedModelIds: number[] = [];

  let axis: AxisReference;
  setLogContext(`Axe / profils (${axesProfils.filename})`);
  if (axesProfils.format === "dxf") {
    axis = buildAxisReferenceFromDxfContent(axesProfils.text as string);
  } else {
    const modelId = openModel(api!, axesProfils.bytes as Uint8Array);
    openedModelIds.push(modelId);
    axis = buildAxisReferenceFromIfcModel(api!, modelId);
  }

  function extractOne(file: FileInput, state: StateKind): { bands: Band[]; samples: WidthSample[]; mode: string } {
    if (file.format === "dxf") {
      const { bands, samples, mode } = extractDxfState(file.text as string, state, axis, gabarit, dxfStepM);
      return { bands, samples, mode };
    }
    const modelId = openModel(api!, file.bytes as Uint8Array);
    openedModelIds.push(modelId);
    const { bands, samples } = extractIfcState(api!, modelId, state, axis);
    return { bands, samples, mode: "ifc" };
  }

  setLogContext(`Existant (${existant.filename})`);
  const existantResult = extractOne(existant, "existant");
  setLogContext(`Projet (${projet.filename})`);
  const projetResult = extractOne(projet, "projet");

  for (const id of openedModelIds) {
    try {
      api!.CloseModel(id);
    } catch {
      // best-effort cleanup
    }
  }

  return {
    axisConfidence: axis.confidence,
    existantMode: existantResult.mode,
    projetMode: projetResult.mode,
    bands: [...existantResult.bands, ...projetResult.bands],
    samples: [...existantResult.samples, ...projetResult.samples],
    axisPoints: axis.axis.points,
    log: getLog(),
  };
}
